services/gitdiff/gitdiff.go

```diff
@@ -555,6 +283,182 @@ func DiffInlineWithHighlightCode(fileName, language, code string) DiffInline {
+			hcd.placeholderTagMap[r] = ""
+		}
+	}
+}
+
+func (hcd *HighlightCodeDiff) diffWithHighlight(filename, language, codeA, codeB string) []diffmatchpatch.Diff {
+	hcd.collectUsedRunes(codeA)
+	hcd.collectUsedRunes(codeB)
+
+	highlightCodeA := highlight.Code(filename, language, codeA)
+	highlightCodeB := highlight.Code(filename, language, codeB)
+
+	highlightCodeA = hcd.convertToPlaceholders(highlightCodeA)
+	highlightCodeB = hcd.convertToPlaceholders(highlightCodeB)
+
+	diffs := diffMatchPatch.DiffMain(highlightCodeA, highlightCodeB, true)
+	diffs = diffMatchPatch.DiffCleanupEfficiency(diffs)
+
+	for i := range diffs {
+		hcd.recoverOneDiff(&diffs[i])
+	}
+	return diffs
+}
+
+func (hcd *HighlightCodeDiff) convertToPlaceholders(htmlCode string) string {
+	var tagStack []string
+	res := strings.Builder{}
+
+	firstRunForLineTags := hcd.lineWrapperTags == nil
+
+	// the standard chroma highlight HTML is "<span class="line [hl]"><span class="cl"> ... </span></span>"
+	for {
+		// find the next HTML tag
+		pos1 := strings.IndexByte(htmlCode, '<')
+		pos2 := strings.IndexByte(htmlCode, '>')
+		if pos1 == -1 || pos2 == -1 || pos2 < pos1 {
+			break
+		}
+		tag := htmlCode[pos1 : pos2+1]
+
+		// write the content before the tag into result string, and consume the tag in the string
+		res.WriteString(htmlCode[:pos1])
+		htmlCode = htmlCode[pos2+1:]
+
+		// the line wrapper tags should be removed before diff
+		if strings.HasPrefix(tag, `<span class="line`) || strings.HasPrefix(tag, `<span class="cl"`) {
+			if firstRunForLineTags {
+				// if this is the first run for converting, save the line wrapper tags for later use, they should be added back
+				hcd.lineWrapperTags = append(hcd.lineWrapperTags, tag)
+			}
+			htmlCode = strings.TrimSuffix(htmlCode, "</span>")
+			continue
+		}
+
+		var tagInMap string
+		if tag[1] == '/' { // for closed tag
+			if len(tagStack) == 0 {
+				break // invalid diff result, no open tag but see close tag
+			}
+			// make sure the closed tag in map is related to the open tag, to make the diff algorithm can match the open/closed tags
+			// the closed tag will be recorded in the map by key "</span><!-- <span the-open> -->" for "<span the-open>"
+			tagInMap = tag + "<!-- " + tagStack[len(tagStack)-1] + "-->"
+			tagStack = tagStack[:len(tagStack)-1]
+		} else { // for open tag
+			tagInMap = tag
+			tagStack = append(tagStack, tag)
+		}
+
+		// remember the placeholder and tag in the map
+		placeholder, ok := hcd.tagPlaceholderMap[tagInMap]
+		if !ok {
+			placeholder = hcd.nextPlaceholder()
+			if placeholder != 0 {
+				hcd.tagPlaceholderMap[tagInMap] = placeholder
+				hcd.placeholderTagMap[placeholder] = tagInMap
+			}
+		}
+
+		if placeholder != 0 {
+			res.WriteRune(placeholder) // use the placeholder to replace the tag
+		} else {
+			res.WriteString(tag) // unfortunately, all private use runes has been exhausted, no more placeholder could be used, so do not convert the tag
```

Comment:
This could/will cause broken tags and hence could result in XSS. If we can't get a placeholder we should abandon highlighting of the diff